src/HealthGPS/energy_balance_model.cpp

```diff
@@ -3,6 +3,17 @@
 
 namespace hgps {
 
+// Rsik factor keys.
+const core::Identifier H_key{"height"};
+const core::Identifier BW_key{"weight"};
+const core::Identifier PAL_key{"physical_activity_level"};
+const core::Identifier F_key{"body_fat"};
+const core::Identifier L_key{"lean_tissue"};
+const core::Identifier ECF_key{"extracellular_fluid"};
+const core::Identifier EI_key{"energy_intake"};
+const core::Identifier EE_key{"energy_expenditure"};
+const core::Identifier CI_key{"carbohydrate"};
```

Comment:
What's the difference in this case? I thought `constexpr` was for things like float initialisation?

Reply:
`const` expressions can be evaluated at runtime, e.g. `const std::string{argv[0]}`, whereas `constexpr` forces it to be evaluated at compile time, latest, a bit more like a C macro with type safety.

Reply:
Is there any practical advantage?

Reply:
Code hygiene? You're basically telling the compiler and reader that what you see in the source is what you get at runtime, and under no circumstances is the value to change. More generally, certain things like static array initialisation require a `constexpr` as the size parameter: `type someArray[myConstExpr];`, else the compiler can't determine how big a stack frame to allocate when entering the function. Besides that, just code hygiene.

Reply:
C++ is odd...

Reply:
Yes, it is. ^^

Reply:
@jamesturner246 If it isn't simple, then don't bother. As you say, not much difference here.

The nice thing about `constexpr` variables in general cf. `const` ones is that you can use them in contexts where you *need* to know the value at compile time, e.g.:

```cpp
constexpr auto i = 4;
std::array<int, i> my_array;
```

This won't work if `i` was just `const`.

The other thing is that you can make functions `constexpr` aware and then you can have basically arbitrary calculations happen at compile time rather than runtime.